src/iocp/port.rs

```diff
@@ -0,0 +1,317 @@
+//! A safe wrapper around the Windows I/O API.
+
+use std::convert::{TryFrom, TryInto};
+use std::fmt;
+use std::io;
+use std::marker::PhantomData;
+use std::mem::MaybeUninit;
+use std::ops::Deref;
+use std::os::windows::io::{AsRawHandle, RawHandle};
+use std::pin::Pin;
+use std::sync::Arc;
+use std::time::Duration;
+
+use windows_sys::Win32::Foundation::{CloseHandle, HANDLE, INVALID_HANDLE_VALUE};
+use windows_sys::Win32::Storage::FileSystem::SetFileCompletionNotificationModes;
+use windows_sys::Win32::System::WindowsProgramming::{FILE_SKIP_SET_EVENT_ON_HANDLE, INFINITE};
+use windows_sys::Win32::System::IO::{
+    CreateIoCompletionPort, GetQueuedCompletionStatusEx, PostQueuedCompletionStatus, OVERLAPPED,
+    OVERLAPPED_ENTRY,
+};
+
+/// A completion block which can be used with I/O completion ports.
+///
+/// # Safety
+///
+/// This must be a valid completion block.
+pub(super) unsafe trait Completion {
+    /// Signal to the completion block that we are about to start an operation.
+    fn try_lock(self: Pin<&Self>) -> bool;
+
+    /// Unlock the completion block.
+    unsafe fn unlock(self: Pin<&Self>);
+}
+
+/// The pointer to a completion block.
+///
+/// # Safety
+///
+/// This must be a valid completion block.
+pub(super) unsafe trait CompletionHandle: Deref + Sized {
+    /// Type of the completion block.
+    type Completion: Completion;
+
+    /// Get a pointer to the completion block.
+    fn get(&self) -> Pin<&Self::Completion>;
+
+    /// Convert this block into a pointer that can be passed as `*mut OVERLAPPED`.
+    fn into_ptr(this: Self) -> *mut OVERLAPPED;
+
+    /// Convert a pointer that was passed as `*mut OVERLAPPED` into a pointer to this block.
+    ///
+    /// # Safety
+    ///
+    /// This must be a valid pointer to a completion block.
+    unsafe fn from_ptr(ptr: *mut OVERLAPPED) -> Self;
+
+    /// Convert to a pointer without losing ownership.
+    fn as_ptr(&self) -> *mut OVERLAPPED;
+}
+
+unsafe impl<'a, T: Completion> CompletionHandle for Pin<&'a T> {
+    type Completion = T;
+
+    fn get(&self) -> Pin<&Self::Completion> {
+        *self
+    }
+
+    fn into_ptr(this: Self) -> *mut OVERLAPPED {
+        unsafe { Pin::into_inner_unchecked(this) as *const T as *mut OVERLAPPED }
+    }
+
+    unsafe fn from_ptr(ptr: *mut OVERLAPPED) -> Self {
+        Pin::new_unchecked(&*(ptr as *const T))
+    }
+
+    fn as_ptr(&self) -> *mut OVERLAPPED {
+        self.get_ref() as *const T as *mut OVERLAPPED
+    }
+}
+
+unsafe impl<T: Completion> CompletionHandle for Pin<Arc<T>> {
+    type Completion = T;
+
+    fn get(&self) -> Pin<&Self::Completion> {
```

Comment:
why is the pointer that this function returns pinned?

Reply:
The pinning ensures that the overlapped object isn't moved while it's involved in an operation.

Reply:
how could it be moved when it is behind an `&` reference?

Reply:
While it's being used in the overlapped operation, there is no reference (or at least, none that the borrow checker is aware of). As a safeguard against undefined behavior, this ensures that it isn't invalidated during the operation.

Reply:
hm, seems a bit brittle, but I'll go with it for now.